demos/device_defender_for_aws/metrics_collector/lwip/metrics_collector.c

```diff
@@ -0,0 +1,209 @@
+    extern union tcp_listen_pcbs_t tcp_listen_pcbs; /* List of all TCP PCBs in LISTEN state */
+
+/**
+ * @brief Get a list of the listening TCP ports.
+ *
+ * This function finds the listening TCP ports by traversing LWIP TCP PCBs list that
+ * in listen state. The header of the list is "tcp_listen_pcbs.listen_pcbs struture".
+ * It can be called with @p pOutTcpPortsArray NULL to get the number of the open TCP ports.
+ *
+ * @param[in] pOutTcpPortsArray The array to write the open TCP ports into. This
+ * can be NULL, if only the number of open ports is needed.
+ * @param[in] tcpPortsArrayLength Length of the pOutTcpPortsArray, if it is not
+ * NULL.
+ * @param[out] pOutNumTcpOpenPorts Number of the open TCP ports.
+ *
+ * @return #MetricsCollectorSuccess if open TCP ports are successfully obtained;
+ * #MetricsCollectorBadParameter if invalid parameters are passed;
+ */
+    MetricsCollectorStatus_t GetOpenTcpPorts( uint16_t * pOutTcpPortsArray,
+                                            uint32_t portsArrayLength,
+                                            uint32_t * pOutNumTcpOpenPorts )
+    {
+        MetricsCollectorStatus_t status = MetricsCollectorSuccess;
+        struct tcp_pcb_listen * pCurrPcb = tcp_listen_pcbs.listen_pcbs;
+        uint16_t pcbCnt = 0;
+
+        if( ( ( pOutTcpPortsArray != NULL ) && ( portsArrayLength == 0 ) ) ||
+            ( pOutNumTcpOpenPorts == NULL ) )
+        {
+            IotLogError( "Invalid parameters. pOutTcpPortsArray: 0x%08x,"
+                         "portsArrayLength: %u, pOutNumTcpOpenPorts: 0x%08x.",
+                         pOutTcpPortsArray,
+                         portsArrayLength,
+                         pOutNumTcpOpenPorts );
+            status = MetricsCollectorBadParameter;
+        }
+
+        if( status == MetricsCollectorSuccess )
+        {
+            LOCK_TCPIP_CORE();
+
+            while( pCurrPcb )
+            {
+                /*  write the ports into pOutTcpPortsArray if not NULL */
+                if( ( pOutTcpPortsArray != NULL ) && ( pcbCnt < portsArrayLength ) )
+                {
+                    pOutTcpPortsArray[ pcbCnt ] = pCurrPcb->local_port;
+                }
+
+                ++pcbCnt;
+                pCurrPcb = pCurrPcb->next;
+            }
+
+            UNLOCK_TCPIP_CORE();
+        }
+
+        if( ( pOutTcpPortsArray != NULL ) && ( pcbCnt > portsArrayLength ) )
+        {
+            IotLogWarn( "The portsArrayLength is not long engough to store all the listening ports" );
+        }
+
+        if( status == MetricsCollectorSuccess )
+        {
+            *pOutNumTcpOpenPorts = pcbCnt;
+        }
+
+        return status;
+    }
+
+
+    MetricsCollectorStatus_t GetOpenUdpPorts( uint16_t * pOutTcpPortsArray,
+                                            uint32_t portsArrayLength,
+                                            uint32_t * pOutNumTcpOpenPorts )
+    {
+        //TODO
```

Comment:
I don't think we should leave TODOs in the code like this. Perhaps a more descriptive comment about how UDP is not current supported?

Reply:
Added UDP support by 753447740470d15789b8bb7c5097c3713facf746